test/e2e/framework/util.go

```diff
@@ -3265,19 +3261,25 @@ func NodeAddresses(nodelist *v1.NodeList, addrType v1.NodeAddressType) []string
 	return hosts
 }
 
-// NodeSSHHosts returns SSH-able host names for all schedulable nodes - this excludes master node.
-// It returns an error if it can't find an external IP for every node, though it still returns all
-// hosts that it found in that case.
+// NodeSSHHosts returns SSH-able host names for all schedulable nodes - this
+// excludes master node. If it can't find any external IPs, it falls back to
+// looking for internal IPs. If it can't find an internal IP for every node it
+// returns an error, though it still returns all hosts that it found in that
+// case.
 func NodeSSHHosts(c clientset.Interface) ([]string, error) {
 	nodelist := waitListSchedulableNodesOrDie(c)
 
-	// TODO(roberthbailey): Use the "preferred" address for the node, once such a thing is defined (#2462).
 	hosts := NodeAddresses(nodelist, v1.NodeExternalIP)
+	// If ExternalIP isn't set, assume the test programs can reach the InternalIP.
+	if len(hosts) == 0 {
```

Comment:
To avoid complications where _some_ nodes have external IPs but not all of them, I just utilized the same fallback logic I saw (and was happy with personally) used: check if there are _no_ external IPs set.

This seems pretty reasonable to me since the setting of the external IPs is provider specific (if I'm not mistaken) so in general they should either all have them or none of them will.

Reply:
can you add this note to a comment in the code as well?